### PR TITLE
Split fuse mount and server launch for flexible use

### PR DESCRIPTION
**Pull request by lkpdn**
When we want a file system to work well with automounting, we may want more flexibility in `fuse.Mount`, i.e. mounting fuse file system in the original process and serving requests in daemonized process(es).
This patch does it in a completely backward-compatible way.

For reference, based on this patch, I cooked up a patch to make `kahing/goofys.git` work well with automounting. (https://github.com/kahing/goofys/pull/369)

Please review, thanks.

### CODE REVIEW

File: mount.go [hunk 1]
```diff
@@ -96,6 +76,68 @@ func Mount(
...
+	// Begin the mounting process, which will continue in the background.
+	dev, err = mount(dir, config, ready)
+	if err != nil {
+		err = fmt.Errorf("mount: %v", err)
+		return
+	}
+	return
+}
+
+func IsMounted(
```
Comment (jacobsa) [inline mount.go]: Why do you need this? Please also add documentation.

File: mount.go [hunk 2]
```diff
@@ -96,6 +76,68 @@ func Mount(
...
 	}
+	return
+}
+
+// FuseMount attempts to mount a file system on the given directory. It can block
+// fuse server setup completion using supplied ready channel.
+func FuseMount(
+	dir string,
+	config *MountConfig,
+	ready chan error) (dev uintptr, err error) {
```
Comment (jacobsa) [inline mount.go]: These use of this channel isn't clear to me. Please elaborate in the documentation what it's for, and when an error may be delivered on the channel versus returned from the function. Please also mark it as a send-only channel to make the intention clear.

File: mount.go [hunk 3]
```diff
@@ -29,49 +29,29 @@ type Server interface {
...
 }
 
-// Mount attempts to mount a file system on the given directory, using the
-// supplied Server to serve connection requests. It blocks until the file
-// system is successfully mounted.
-func Mount(
-	dir string,
+// FuseServe attempts to start the supplied fuse server with the file discriptor
+// to poll against. It blocks until the file system is successfully mounted.
+func FuseServe(
```
Comment (jacobsa) [inline mount.go]: Could you please describe in the documentation for this and the function below when you would want to use them versus just using Mount? Having three entry points is confusing; we should at least have the more complex ones give a pointer to the simpler ones.

File: mount.go (hunk 1 above)
Comment (lkpdn) [reply mount.go]: This is just a util func and some still do not need this. Now I'm wondering this should be implemented in client side, not in this jacobsa/fuse.git. Perhaps I'll remove these.

File: mount.go (hunk 2 above)
Comment (lkpdn) [reply mount.go]: Okay I'll add a comment above the func definition, explaining what it's for + when error delivered.

> Please also mark it as a send-only channel to make the intention clear.

I'll push a commit later. Thanks.

File: mount.go (hunk 3 above)
Comment (lkpdn) [reply mount.go]: I'll add a comment. Thanks.

**Comment by lkpdn** [comment]
Thanks for your quick response.

> Hi, could you please give some more info on what this is for? For example, could you summarize how you intend to use this? I read the goofyfs pull request description, but it doesn't make any sense to me (perhaps because I'm not familiar with goofyfs and automount). Some intermediate level of detail would be helpful.

Could you please wait for a while until sufficient discussion having been done on the goofys PR? Then I'll mention you here on this PR for notification.
